### PR TITLE
Go AST blogpost, part 2

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
Added new blog post and support for 'series' (it adds a previous/next links in the sidebar)

### CODE REVIEW

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+series: "AST"
+date: 2020-03-15T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+---
+
+## Welcome!
+
+This is a second part of our AST blog post series, expanding on the subject of utilizing Go AST libraries libraries to automate and improve your workflow.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
This is the second part of our AST blog post series, expanding on the subject of utilizing Go AST libraries to automate and improve your workflow.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+In this post I'll discuss a rather common problem that comes up while working with Go code and the way we've solved it by sprinkling a little bit of AST magic dust. Let's dive in.
+
+## Problem: A struct with no interface
+
+Let's say you are working on a large code base that was not built with interfaces in mind, meaning, there are `structs` and methods attached to that `struct`, but there is not `interface` describing it. This is a perfectly valid approach when no you don't need to mock/stub the method implementations provided by that struct, or when there isn't more than one implementation of the same 'contract'. When these things **are** needed - we need to provide an `interface`.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Let's say you are working on a large code base that was not built with interfaces in mind, meaning, there are `structs` and methods attached to those `structs`, but there is no `interface` describing it. This is a perfectly valid approach when you don't need to mock/stub the method implementations provided by that `struct`, or when there isn't more than one implementation of the same 'contract'. When these things **are** needed - we need to provide an `interface`.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	a := Animal{Legs: 4}
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` structs, we would have to define it as:
+```go
+list := []interface{}{p,a}
+```
+But this way we loose the type information of the list elements.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
But this way we lose the type information of the list elements.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+						if r, rok := fun.Recv.List[0].Type.(*ast.StarExpr); rok && r.X.(*ast.Ident).Name == structName {
+							// we found it!
+						}
+					}
+				}
+
+			}
+			return true
+		})
+	}
```
Comment (hanzei) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
fset := token.NewFileSet()
// 1. scan source code folder
pkgs, err := parser.ParseDir(fset, folder, nil, parser.AllErrors|parser.ParseComments)
if err != nil {
	log.Fatalf("Unable to parse %s folder", folder)
}
// 2. find the required package by name
var appPkg *ast.Package
for _, pkg := range pkgs {
	if pkg.Name == pkgName {
		appPkg = pkg
		break
	}
}
if appPkg == nil {
	log.Fatalf("Unable to find package %s", pkgName)
}

// 3. find all methods that are bound to the specific struct
for _, file := range appPkg.Files {
	ast.Inspect(file, func(n ast.Node) bool {
		if fun, ok := n.(*ast.FuncDecl); ok {
			// 4. Validate that method is exported and has a receiver
			if fun.Name.IsExported() && fun.Recv != nil && len(fun.Recv.List) == 1 {
					// 5. Check that the receiver is actually the struct we want
					if r, rok := fun.Recv.List[0].Type.(*ast.StarExpr); rok && r.X.(*ast.Ident).Name == structName {
						// we found it!
					}
				}
			}

		}
		return true
	})
}
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+```go
+	sort.Strings(funcs)
+	out := bytes.NewBufferString("")
+
+	t := template.Must(template.New("").Parse(outputTemplate))
+	err = t.Execute(out, map[string]interface{}{
+		"Content": strings.Join(funcs, "\n"),
+		"Name":    ifName,
+		"Package": pkgName,
+	})
```
Comment (hanzei) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
sort.Strings(funcs)
out := bytes.NewBufferString("")

t := template.Must(template.New("").Parse(outputTemplate))
err = t.Execute(out, map[string]interface{}{
	"Content": strings.Join(funcs, "\n"),
	"Name":    ifName,
	"Package": pkgName,
})
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func main() {
+	p := Person{Name: "Fred"}
+	a := Animal{Legs: 4}
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` structs, we would have to define it as:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` `structs`, we would have to define it as:
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+We are almost done! Our interface file is ready, but it's missing a crucial part - imports. Luckily, there is a package for that™ - https://pkg.go.dev/golang.org/x/tools/imports
+
+Let give it whirl:
+```go
+	formatted, err := imports.Process(outputFile, out.Bytes(), &imports.Options{Comments: true})
+	if err != nil {
+		log.Panic(err)
+	}
+	err = ioutil.WriteFile(outputFile, formatted, 0644)
```
Comment (hanzei) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
formatted, err := imports.Process(outputFile, out.Bytes(), &imports.Options{Comments: true})
if err != nil {
	log.Panic(err)
}
err = ioutil.WriteFile(outputFile, formatted, 0644)
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` structs, we would have to define it as:
+```go
+list := []interface{}{p,a}
+```
+But this way we loose the type information of the list elements.
+This is where interfaces come in. Since both `Person` and `Animal` _implement_ a method with the same signature, we can extract that signature into an interface and use it for storing items in a list:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
This is where interfaces come in. Since both `Person` and `Animal` _implement_ a method with the same signature, we can extract that signature into an `interface` and use it for storing items in a list:
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+```go
+type interface Hello {
+	Hello() string
+}
+// ...
+list := []Hello{p,a}
+fmt.Printf("Person: [%v] Animal: [%v]\n", list[0].Hello(), list[1].Hello())
+```
+
+Awesome. Now let's say the struct you are extracting the interface from is a big on. A really big one. With lots and lots of methods spread out in different `.go` files. Creating such interface manually would be very laborious.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Awesome. Now let's say the `struct` you are extracting the interface from is a big one. A really big one. With lots and lots of methods spread out in different `.go` files. Creating such an interface manually would be very laborious.  
```

File: site/layouts/partials/series_link.html [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    {{ end }}
+    {{ range where .Site.RegularPages.ByDate.Reverse "Params.series" .Params.series }} 
+        {{ $.Scratch.Set "first_link" .Permalink }}
+        {{ if lt $.Date.Unix .Date.Unix }}
+            {{ $.Scratch.Set "next_link" .Permalink }}
+            {{ $.Scratch.Set "next_title" .Title }}
+        {{ end }}
+    {{ end }}
+    {{ if or ($.Scratch.Get "next_link") ($.Scratch.Get "prev_link") }} 
+        <hr/>
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: This line looks weird, as it doesn't align the line of the left. What was the reason for adding it?
Screenshot from 2020-03-13 11-16-53

File: site/layouts/partials/series_link.html [hunk 11]
```diff
@@ -0,0 +1,40 @@
+{{ if .Params.series }}
+    {{ $.Scratch.Add "cur_page_num" 1 }}
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: Why is a normal variable not enough and you need need `Scratch`?

File: site/layouts/partials/series_link.html [hunk 12]
```diff
@@ -0,0 +1,40 @@
+{{ if .Params.series }}
+    {{ $.Scratch.Add "cur_page_num" 1 }} 
+    {{ $.Scratch.Add "total_page_num" 0 }} 
+    {{ range where .Site.RegularPages.ByDate "Params.series" .Params.series }} 
+        {{ $.Scratch.Add "total_page_num" 1 }} 
+        {{ if gt $.Date.Unix .Date.Unix }} 
+            {{ $.Scratch.Add "cur_page_num" 1 }} 
+            {{ $.Scratch.Set "prev_link" .Permalink }}
+            {{ $.Scratch.Set "prev_title" .Title }}
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: Nit: Are the white spaces at the end of the line needed?

File: site/layouts/partials/series_link.html [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        {{ end }}
+    {{ end }}
+    {{ if or ($.Scratch.Get "next_link") ($.Scratch.Get "prev_link") }} 
+        <hr/>
+        <p>Part {{ $.Scratch.Get "cur_page_num" }} of {{ $.Scratch.Get "total_page_num" }}
+            in the <b>{{ .Params.series }}</b> series.</p>
+        <p>
+        {{ if $.Scratch.Get "prev_link" }}
+            {{ if ne ($.Scratch.Get "prev_link") ($.Scratch.Get "first_link") }}
+                <a href="{{ $.Scratch.Get "first_link" }}"><i class="fa fa-angle-left"></i>Series Start<i class="fa fa-angle-right"></i></a>
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: I'm not sure if a `Series Start` link is really needed. Why did you add it?

File: site/layouts/partials/series_link.html [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+                    |
+            {{ end }}
+            {{ if $.Scratch.Get "prev_link" }}
+                <a href="{{ $.Scratch.Get "prev_link" }}"><i class="fa fa-angle-double-left"></i>{{ $.Scratch.Get "prev_title" }}</a>
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: If there is both a prev and next page, it starts to look a bit crowded. What about omitting the title for just `Part X`?
Screenshot from 2020-03-13 11-22-38

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md (hunk 9 above)
Comment (hanzei) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]: Why does the interface have to contain _all_ methods of a struct? In general the user of the interface only cares about the methods that are needed. Interfaces are there to allow abstraction, but if every public method is on the interface, there is not abstraction.

**Comment by reflog** [comment]
@hanzei - addressed your comments

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+1. Scan the source code for all methods implemented on a specific `struct`
+1. Collect all those methods (and their comments, for clarity)
+1. Create a new file that will contain an interface with all the collected methods inside
+1. ...
+1. Profit?
+
+### Scanning the source code for all struct's methods
+
+Here's a short piece of code that scans a folder of go source code to first finds a package by it's name and then searches for all methods that are bound to the struct we are interested in.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Here's a short piece of code that scans a folder of `go` source code to first find a package by name and then search for all methods that are bound to the `struct` we're interested in.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+			}
+			return true
+		})
+	}
+```
+
+Steps 1 and 2 are pretty straightforward, the interesting bits start at step 3: For each file in the package, we execute `ast.Inspect` to get all the AST nodes. For every node that is actually a function (checked by `n.(*ast.FuncDecl)`), we check if that function is:  
+* Exported (we are not interested in private methods)
+* Has a receiver (it's bound to a struct)
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
* Has a receiver (it's bound to a `struct`)
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+* Receiver's type matches the struct we are interested in
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
* Receiver's type matches the `struct` we are interested in
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+}
+// ...
+list := []Hello{p,a}
+fmt.Printf("Person: [%v] Animal: [%v]\n", list[0].Hello(), list[1].Hello())
+```
+
+Awesome. Now let's say the struct you are extracting the interface from is a big on. A really big one. With lots and lots of methods spread out in different `.go` files. Creating such interface manually would be very laborious.  
+This problem is itching to get an AST treatment. Let's get to it!
+
+## AST To The Rescue!
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
## AST to the rescue!
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+* Has a receiver (it's bound to a struct)
+* Receiver's type matches the struct we are interested in
+
+### Collecting functions and their comments
+
+Now that we can get a `*ast.FuncDecl` (let's call it `fun`) for each function, we can collect all the information about it and reconstruct it:
+
+1. Name: `fun.Name.Name`
+2. Comments: `fun.Doc.List`
+3. Parameters: fun.Type.Params.Lis
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
3. Parameters: `fun.Type.Params.Lis`
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+4. Return values: fun.Type.Results.List
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
4. Return values: `fun.Type.Results.List`
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+	t := template.Must(template.New("").Parse(outputTemplate))
+	err = t.Execute(out, map[string]interface{}{
+		"Content": strings.Join(funcs, "\n"),
+		"Name":    ifName,
+		"Package": pkgName,
+	})
+```	
+
+We are almost done! Our interface file is ready, but it's missing a crucial part - imports. Luckily, there is a package for that™ - https://pkg.go.dev/golang.org/x/tools/imports
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
We're almost done! Our `interface` file is ready, but it's missing a crucial part - `imports`. Luckily, there is a package for that™ - https://pkg.go.dev/golang.org/x/tools/imports.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+
+Let give it whirl:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Let's give it whirl:
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+		log.Panic(err)
+	}
+	err = ioutil.WriteFile(outputFile, formatted, 0644)
+```	
+
+Voila! You have an interface file that contains all the methods implemented on the `struct`.
+
+## struct2interface
+
+The process I've described is a rather generic one, so to make it fully repeatable, I've extracted it into a separate CLI utility called `struct2interface`. It can be found at https://github.com/reflog/struct2interface
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
The process I've described is a rather generic one, so to make it fully repeatable, I've extracted it into a separate CLI utility called `struct2interface`. It can be found at https://github.com/reflog/struct2interface.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+
+
+## Conclusion
+
+Once again, AST came to the rescue and saved us lots and lots of manual labor. Hurray! In the next post of these series, I'll describe how we combined our `layers` approach with AST to create a clean `opentracing` instrumentation we've started in the first part of these series.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Once again, AST came to the rescue and saved us lots and lots of manual labor. Hurray! In the next post of this series, I'll describe how we combined our `layers` approach with AST to create a clean `opentracing` instrumentation we've started in the first part of these series.
```

**Comment by reflog** [comment]
@justinegeffen - thanks for all the corrections! You are right, I was inconsistent with quoting the terms struct/interface, I'll make another pass and make sure they are all quoted.

**Comment by reflog** [comment]
@hanzei - I've added a note: https://github.com/mattermost/mattermost-developer-documentation/pull/558/files#diff-a1cd618ecc3c531716973490c5bfb9c6R77 
What do you think?

**Comment by justinegeffen** [comment]
> @justinegeffen - thanks for all the corrections! You are right, I was inconsistent with quoting the terms struct/interface, I'll make another pass and make sure they are all quoted.

No worries at all! I wasn't sure whether my changes to struct/interface etc were contextually correct so thanks for giving it another pass. :)

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+
+## Welcome!
+
+This is the second part of our AST blog post series, expanding on the subject of utilizing Go AST libraries to automate and improve your workflow.
+
+In this post I'll discuss a rather common problem that comes up while working with Go code and the way we've solved it by sprinkling a little bit of AST magic dust. Let's dive in.
+
+## Problem: A `struct` with no `interface`
+
+Let's say you are working on a large code base that was not built with `interfaces` in mind, meaning, there are `structs` and methods attached to those `structs`, but there is no `interface` describing it. This is a perfectly valid approach when you don't need to mock/stub the method implementations provided by that `struct`, or when there isn't more than one implementation of the same 'contract'. When these things **are** needed - we need to provide an `interface`.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Let's say you are working on a large code base that was not built with `interfaces` in mind, meaning, there are `structs` and methods attached to those `structs`, but there is no `interface` describing it. This is a perfectly valid approach when you don't need to mock/stub the method implementations provided by that `struct`, or there's only one implementation of the same 'contract'. 

However, when these things are required we need to provide an `interface`.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func main() {
+	p := Person{Name: "Fred"}
+	a := Animal{Legs: 4}
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` `structs`, we would have to define it as:
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of both the `Person` and `Animal` `structs`, we would have to define it as:
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+
+Let's break down the task at hand into smaller, digestable parts:
+
+1. Scan the source code for all methods implemented on a specific `struct`
+1. Collect all those methods (and their comments, for clarity)
+1. Create a new file that will contain an `interface` with all the collected methods inside
+1. ...
+1. Profit?
+
+> Note: usually the `interface` doesn't contain **all** of the `struct` methods, but we'll leave the topics of abstraction and better code organization for another blog post
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
**Note:** Usually the `interface` doesn't contain all of the `struct` methods, but we'll leave the topics of abstraction and better code organization for another blog post.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+
+
+### Scanning the source code for all `struct`'s methods
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
### Scanning the source code for all `struct` methods
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+
+Here's a short piece of code that scans a folder of `go` source code to first find a package by name and then search for all methods that are bound to the `struct` we're interested in.
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
Here's a short piece of code that scans a folder of `.go` source code to first find a package by name and then search for all methods that are bound to the `struct` we're interested in.
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+			}
+
+		}
+		return true
+	})
+}
+```
+
+Steps 1 and 2 are pretty straightforward, the interesting bits start at step 3: For each file in the package, we execute `ast.Inspect` to get all the AST nodes. For every node that is actually a function (checked by `n.(*ast.FuncDecl)`), we check if that function is:  
+* Exported (we are not interested in private methods)
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion

* Exported (we are not interested in private methods)
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+	"Content": strings.Join(funcs, "\n"),
+	"Name":    ifName,
+	"Package": pkgName,
+})
+```	
+
+We're almost done! Our `interface` file is ready, but it's missing a crucial part - `imports`. Luckily, there is a package for that™ - https://pkg.go.dev/golang.org/x/tools/imports.
+
+Let's give it whirl:
+```go
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion

```go
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md [hunk 32]
```diff
@@ -0,0 +1,186 @@
...
+formatted, err := imports.Process(outputFile, out.Bytes(), &imports.Options{Comments: true})
+if err != nil {
+	log.Panic(err)
+}
+err = ioutil.WriteFile(outputFile, formatted, 0644)
+```	
+
+Voila! You have an `interface` file that contains all the methods implemented on the `struct`.
+
+## struct2interface
```
Comment (justinegeffen) [inline site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]:
```suggestion
## `struct2interface`
```

File: site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md (hunk 28 above)
Comment (justinegeffen) [reply site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md]: Not sure whether this suggestion works - are we referring to the plural or a specific struct's method?

File: site/layouts/blog/single.html [hunk 33]
```diff
@@ -52,6 +52,8 @@
             <!-- Sidebar -->
             <div class="col-md-3 doc-content tags-sidebar">
                 <div class="well well-sm">
+                    {{ partial "series_link.html" . }}
+
```
Comment (hanzei) [inline site/layouts/blog/single.html]:
```suggestion
                    {{ partial "series_link.html" . }}
```

File: site/layouts/partials/series_link.html [hunk 34]
```diff
@@ -0,0 +1,39 @@
+{{ if .Params.series }}
+    {{ $cur_page_num := 1 }}
+    {{ $total_page_num := 0 }}
+    {{ $prev_link := "" }}
+    {{ $prev_title := "" }}
+    {{ $next_link := "" }}
+    {{ $next_title := "" }}
+    {{ $first_link := "" }}
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: Do you still need `first_link`?

File: site/layouts/partials/series_link.html [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+    {{ range where .Site.RegularPages.ByDate.Reverse "Params.series" .Params.series }}
+        {{ $first_link = .Permalink }}
+        {{ if lt $.Date.Unix .Date.Unix }}
+            {{ $next_link = .Permalink }}
+            {{ $next_title = .Title }}
+        {{ end }}
+    {{ end }}
+    {{ if or $next_link $prev_link }}
+        <p>Part {{ $cur_page_num }} of {{ $total_page_num }}
+            in the <b>{{ .Params.series }}</b> series.</p>
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: Nit: Make space it a bit different
```suggestion
        <p>
        	Part {{ $cur_page_num }} of {{ $total_page_num }} in the <b>{{ .Params.series }}</b> series.
        </p>
```

File: site/layouts/partials/series_link.html [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+            <a href="{{ $prev_link }}"><i class="fa fa-angle-double-left"></i>&nbsp;Part {{ sub $cur_page_num 1 }}</a>
+        {{ end }}
+        {{ if and $next_link $prev_link }}
+                |
+        {{ end }}
+        {{ if $next_link }}
+            <a href="{{ $next_link }}">Part {{ add $cur_page_num 1 }}&nbsp;<i class="fa fa-angle-double-right"></i></a></p>
+        {{ end }}
+    {{ end }}
+{{ end }}
```
Comment (hanzei) [inline site/layouts/partials/series_link.html]: Nit: Add a line break
